File: public-study/week15/17144/dckat.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int a[50][50];  // 입력 정보
int b[50][50];  // 확산된 미세먼지 변화량 저장
int dx[] = {0, -1, 0, 1};       // 이동 방향
int dy[] = {1, 0, -1, 0};       // 이동 방향
int r, c, t;    // 입력값
int sx, sy;     // 공기 청정기 위치 (윗행 기준)

void input() {
    cin >> r >> c >> t;

    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            cin >> a[i][j];

            // 공기청정기 위치
            if (a[i][j] == -1) {
                sx = i;
                sy = j;
            }
        }
    }
}

// 범위 확인
bool isBound(int x, int y) {
    if (0 <= x && x < r && 0 <= y && y < c) {
        return true;
    }
    return false;
}

// 확산 가능여부 확인
bool canMove(int x, int y) {
    // 영역 내에 존재 및 공기청정기가 없는 경우
    if (isBound(x, y) && a[x][y] != -1) {
        return true;
    }
    return false;
}

void go(int sX, int sY, int dir) {
    int pre = 0;
    int x = sX;
    int y = sY + 1;
    int k = 0;

    while (true) {
        if (x == sX && y == sY) {
            break;
        }
        swap(pre, a[x][y]);

        x += dx[k];
        y += dy[k];
        
        // 경계를 넘어간 경우 방향 전환
        // 위: 반시계 방향, 아래: 시계 방향
        if (!isBound(x, y)) {
            x -= dx[k];
            y -= dy[k];
            k = (k+dir) % 4;

            x += dx[k];
            y += dy[k];
        }
        
    }
}

void solution() {
    while (t--) {
        // 1번 케이스: 미세먼지 확산
        for (int x = 0; x < r; x++) {
            for (int y = 0; y < c; y++) {
                if (a[x][y] > 0) {
                    int cnt = 0;
                    int val = a[x][y] / 5;

                    // 확산 가능한 미세먼지 영역 찾고 값 변경
                    for (int k = 0; k < 4; k++) {
                        int nx = x + dx[k];
                        int ny = y + dy[k];

                        if (canMove(nx, ny)) {
                            b[nx][ny] += val;
                            cnt++;
                        }
                    }
                    a[x][y] -= cnt*val;
                }
            }
        }

        // 확산된 양만큼 미세먼지 현황 변경
        for (int x = 0; x < r; x++) {
            for (int y = 0; y < c; y++) {
                a[x][y] += b[x][y];
                b[x][y] = 0;
            }
        }

        // 공기청정기 이동 (위.아래 각각 공기 순환)
        go(sx-1, sy, 1);
        go(sx, sy, 3);
    }

    int answer = 0;
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            if (a[i][j] > 0) {
                answer += a[i][j];
            }
        }
    }
    cout << answer << '\n';
}

int main() {
    input();
    solution();
}

File: public-study/week18/1941/dckat.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    private static int answer = 0;
    private static char[][] board = new char[5][5];
    private static boolean selected[][] = new boolean[5][5];

    private static int[] dx = {-1, 1, 0, 0};
    private static int[] dy = {0, 0, -1, 1};

    private static void input() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int i = 0; i < 5; i++) {
            char[] arr = br.readLine().toCharArray();
            for (int j = 0; j < 5; j++) {
                board[i][j] = arr[j];
            }
        }
    }

    // 범위 확인
    private static boolean isBound(int x, int y) {
        if (0 <= x && x < 5 && 0 <= y && y < 5) {
            return true;
        }
        return false;
    }

    // 다음으로 이동가능한지 확인
    private static boolean canMove(int nx, int ny) {
        if (isBound(nx, ny)) {
            // 다음으로 이동할 곳이 선택된 후보인 경우
            if (selected[nx][ny]) {
                return true;
            }
        }
        return false;
    }

    private static boolean bfs(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[5][5];

        q.add(new int[]{x, y});
        visited[x][y] = true;

        int cnt = 1;
        while(!q.isEmpty()) {
            int[] cur = q.poll();

            // 상하좌우 인접 탐색
            for (int k = 0; k < 4; k++) {
                int nx = cur[0] + dx[k];
                int ny = cur[1] + dy[k];

                // 다음으로 이동가능하고 방문하지 않은 경우 큐에 추가
                if (canMove(nx, ny) && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    q.add(new int[]{nx, ny});
                    cnt++;
                }
            }
        }
        return cnt == 7;
    }

    private static void solution(int idx, int sCnt, int yCnt, int total) {
        // 임도연파가 4명 이상인 경우 수행하지 않고 0을 반환하여 종료
        if (yCnt >= 4) {
            return;
        }

        // 7명을 채운 경우
        if (total == 7) {
            // 7명이 모두 인접해 있는지 확인
            int curIdx = idx - 1;
            if (bfs(curIdx/5, curIdx%5)) {
                answer++;
            }
            return;
        }

        // 모든 조합을 탐색
        for (int i = idx; i < 25; i++) {
            int x = i/5;
            int y = i%5;
            selected[x][y] = true;

            // 이다솜파가 선택된 경우
            if (board[x][y] == 'S') {
                solution(i+1, sCnt+1, yCnt, total+1);
            }
            // 임도연파가 선택된 경우
            else {
                solution(i+1, sCnt, yCnt+1, total+1);
            }
            selected[x][y] = false;
        }
    }

    public static void main(String[] args) throws Exception {
        input();
        solution(0,0, 0, 0);
        System.out.println(answer);
    }
}